nexusclient/client.go

```diff
@@ -0,0 +1,725 @@
+		default:
+			return nil, c.newUnexpectedResponseError(fmt.Sprintf("unexpected operation state: %s", state), response)
+		}
+	case nexusapi.StatusOperationFailed:
+		state, err := c.getUnsuccessfulStateFromHeader(response)
+		if err != nil {
+			return nil, err
+		}
+		failure, err := c.failureFromResponse(response)
+		if err != nil {
+			return nil, err
+		}
+		return nil, &nexusapi.UnsuccessfulOperationError{
+			State:   state,
+			Failure: failure,
+		}
+	default:
+		return nil, c.newUnexpectedResponseError(fmt.Sprintf("unexpected response status: %q", response.Status), response)
+	}
+}
+
+// Options for [nexusclient.OperationHandle.Cancel].
+type CancelOptions struct {
+	// Header to attach to the HTTP request. Optional.
+	Header http.Header
+}
+
+// Cancel requests to cancel an asynchronous operation.
+//
+// Cancelation is asynchronous and may be not be respected by the operation's implementation.
+func (h *OperationHandle) Cancel(ctx context.Context, options CancelOptions) error {
+	if h.id == "" {
+		return ErrHandleForSyncOperation
+	}
+
+	return h.client.cancelOperation(ctx, cancelOperationRequest{
+		Operation:   h.operation,
+		OperationID: h.id,
+		Header:      options.Header,
+	})
+}
+
+type cancelOperationRequest struct {
+	Operation   string
+	OperationID string
+	Header      http.Header
+}
+
+func (c *Client) cancelOperation(ctx context.Context, request cancelOperationRequest) error {
+	url, err := c.joinURL(request.Operation, request.OperationID, "cancel")
+	if err != nil {
+		return err
+	}
+	httpReq, err := http.NewRequestWithContext(ctx, "POST", url.String(), nil)
+	if err != nil {
+		return err
+	}
+	if request.Header != nil {
+		httpReq.Header = request.Header.Clone()
+	}
+
+	httpReq.Header.Set(headerUserAgent, UserAgent)
+	response, err := c.Options.HTTPClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+	if response.StatusCode != http.StatusAccepted {
+		return c.newUnexpectedResponseError(fmt.Sprintf("unexpected response status: %q", response.Status), response)
+	}
+	return nil
+}
+
+// Input for [nexusclient.Client.StartOperation].
+type StartOperationRequest struct {
+	// Name of the operation to start.
+	Operation string
+	// Callback URL to provide to the handle for receiving async operation completions. Optional.
+	CallbackURL string
+	// Request ID that may be used by the server handler to dedupe this start request.
+	// By default a v4 UUID will be generated by the client.
+	RequestID string
+	// Header to attach to the HTTP request. Optional.
+	Header http.Header
+	// Body of the operation request.
+	// If it is an [io.Closer], the body will be automatically closed by the client.
+	Body io.Reader
+}
+
+// NewJSONStartOperationRequest is shorthand for creating a StartOperationRequest with a JSON body.
+// Marhsals the provided value to JSON using [nexusapi.DefaultMarshaler].
+func NewJSONStartOperationRequest(operation string, v any) (request StartOperationRequest, err error) {
```

Comment:
```suggestion
func NewStartOperationRequest(operation string, body any) (request StartOperationRequest, err error) {
```

I think we can assume the body is always JSON on non-nil `body` for most uses

Reply:
Hmm.. I find this a bit confusing and would rather be explicit.

Reply:
It's very normal in Go for the explicit struct instantiation to be the more advanced way and the function to be the simpler way with accepted defaults.